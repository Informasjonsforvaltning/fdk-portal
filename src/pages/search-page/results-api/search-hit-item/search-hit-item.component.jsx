import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';

import localization from '../../../../lib/localization';
import './search-hit-item.scss';
import { getTranslateText } from '../../../../lib/translateText';
import { SearchHitHeader } from '../../../../components/search-hit-header/search-hit-header.component';
import { AlertMessage } from '../../../../components/alert-message/alert-message.component';
import {
  iconIsFree,
  iconIsNotFree,
  iconIsNotOpenAccess,
  iconIsNotOpenLicense,
  iconIsOpenAccess,
  iconIsOpenLicense
} from '../../../../components/api-icons';
import { getFirstLineOfText } from '../../../../lib/stringUtils';

const renderHeaderLink = (item, publisher, publishers, referenceData) => {
  if (!item) {
    return null;
  }
  const { title } = item;
  const link = `/apis/${encodeURIComponent(item.id)}`;
  return (
    <header>
      <SearchHitHeader
        tag="h2"
        title={getTranslateText(title)}
        titleLink={link}
        publisherLabel={`${localization.provider}:`}
        publisher={publisher}
        publisherItems={publishers}
        nationalComponent={item.nationalComponent}
        statusCode={item.statusCode}
        referenceData={referenceData}
        detialsPage={false}
      />
    </header>
  );
};

const renderDescription = description => {
  if (!description) {
    return null;
  }

  return (
    <div className="fdk-text-size-medium">
      <div className="uu-invisible" aria-hidden="false">
        Beskrivelse av api
      </div>
      {getFirstLineOfText(getTranslateText(description))}
    </div>
  );
};

const renderExpiredOrDeprecatedVersion = item => {
  const statusCode = _.get(item, 'statusCode');
  const deprecationInfoExpirationDate = _.get(
    item,
    'deprecationInfoExpirationDate'
  );
  const deprecationInfoReplacedWithUrl = _.get(
    item,
    'deprecationInfoReplacedWithUrl'
  );
  if (statusCode === 'REMOVED') {
    return (
      <AlertMessage type="info">
        <span>{localization.statusRemoved} </span>
        {deprecationInfoReplacedWithUrl && (
          <a href={deprecationInfoReplacedWithUrl}>
            {localization.statusReplacedUrl}
          </a>
        )}
      </AlertMessage>
    );
  }
  if (statusCode === 'DEPRECATED') {
    return (
      <AlertMessage type="info">
        <span>
          {deprecationInfoExpirationDate &&
            localization.formatString(
              localization.statusDeprecated,
              localization.during,
              deprecationInfoExpirationDate.substring(0, 4)
            )}
          {!deprecationInfoExpirationDate &&
            localization.formatString(localization.statusDeprecated, '', '')}
        </span>

        {deprecationInfoReplacedWithUrl && (
          <a href={deprecationInfoReplacedWithUrl}>
            {localization.statusReplacedUrl}
          </a>
        )}
      </AlertMessage>
    );
  }
  return null;
};

const renderFormat = formats => {
  if (!formats || formats.length === 0) {
    return null;
  }

  const formatItems = formats =>
    formats.map((item, index) => (
      <span key={index}>
        {index > 0 ? ', ' : ''}
        {item}
      </span>
    ));

  return (
    <div className="mb-2">
      <strong>{localization.format}:&nbsp;</strong>
      {formatItems(formats)}
    </div>
  );
};

export const SearchHitItem = ({
  item,
  fadeInCounter,
  publishers,
  referenceData
}) => {
  const searchHitClass = cx('search-hit', {
    'fade-in-200': fadeInCounter === 0,
    'fade-in-300': fadeInCounter === 1,
    'fade-in-400': fadeInCounter === 2
  });

  const { isFree, isOpenAccess, isOpenLicense } = item;

  return (
    <article className={searchHitClass}>
      <span className="uu-invisible" aria-hidden="false">
        SÃ¸ketreff.
      </span>

      {renderHeaderLink(
        item,
        _.get(item, 'publisher'),
        publishers,
        referenceData
      )}

      {renderExpiredOrDeprecatedVersion(item)}

      {renderDescription(_.get(item, 'description'))}

      {renderFormat(_.get(item, 'formats'))}

      <div className="access-icons mt-2 d-flex justify-content-between">
        {isFree === true && iconIsFree()}
        {isFree === false && iconIsNotFree()}
        {isOpenAccess === true && iconIsOpenAccess()}
        {isOpenAccess === false && iconIsNotOpenAccess()}
        {isOpenLicense === true && iconIsOpenLicense()}
        {isOpenLicense === false && iconIsNotOpenLicense()}
      </div>
    </article>
  );
};

SearchHitItem.defaultProps = {
  referenceData: null,
  item: {},
  publishers: null
};

SearchHitItem.propTypes = {
  item: PropTypes.shape({}),
  publishers: PropTypes.object,
  referenceData: PropTypes.object
};
