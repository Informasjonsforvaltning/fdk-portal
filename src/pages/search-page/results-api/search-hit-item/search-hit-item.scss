@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.search-hit {
  background-color: $color-white;
  border-radius: 5px;
  color: $color-neutral-darkest;
  padding: 1.5em 2em;
  margin-bottom: 1em;
  .search-hit__version {
    background-color: $color-primary-lighter;
    border-radius: 2px;
  }
}
[class^="icon2-"] {
    font-size: 34px;
    vertical-align: -9px;
    padding-right: 10px;
}
