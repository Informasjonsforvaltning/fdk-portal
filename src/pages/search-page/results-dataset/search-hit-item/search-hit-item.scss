@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.fdk-a-search-hit {
  text-decoration: none;
}

.theme-fdk {
  .fdk-icon-public {
    background-image: url("/img/icon-dataset-public-md.svg");
  }
}

.theme-nap {
  .fdk-icon-public {
    background-image: url("/img/icon-dataset-public-nap-md.svg");
  }
}

.fdk-icon-public {
  background-image: url("/img/icon-dataset-public-md.svg");
  width: 2em; 
  height: 2em;
}

.fdk-icon-non-public {
  background-image: url("/img/icon-dataset-not-public-md.svg");
  width: 2em; 
  height: 2em;
}

.fdk-icon-restricted {
  background-image: url("/img/icon-dataset-restricted-md.svg");
  width: 2em; 
  height: 2em;
}

.fdk-color-restricted {
  color: $color-warning-darker !important;
}