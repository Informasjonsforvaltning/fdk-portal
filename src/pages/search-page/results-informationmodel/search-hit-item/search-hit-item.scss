@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.search-hit {
  background-color: $color-white;
  border-radius: 5px;
  color: $color-neutral-darkest;
  padding: 1.5em 2em;
  margin-bottom: 1em;
}
