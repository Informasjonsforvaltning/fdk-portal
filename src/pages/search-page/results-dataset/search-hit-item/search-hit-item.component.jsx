import React from 'react';
import PropTypes from 'prop-types';
import * as _ from 'lodash';

import localization from '../../../../lib/localization';
import { getTranslateText } from '../../../../lib/translateText';
import { SearchHitHeader } from '../../../../components/search-hit-header/search-hit-header.component';
import './search-hit-item.scss';
import {
  getReferenceDataByUri,
  REFERENCEDATA_PATH_DISTRIBUTIONTYPE
} from '../../../../redux/modules/referenceData';

const renderFormats = (source, referenceData) => {
  const { distribution } = source;

  const children = distributions => {
    const nodes = [];
    distributions.forEach(item => {
      let formatString;
      const { format } = item;
      let { type } = item;
      if (
        type &&
        (type !== 'API' && type !== 'Feed' && type !== 'Nedlastbar fil')
      ) {
        const distributionType = getReferenceDataByUri(
          referenceData,
          REFERENCEDATA_PATH_DISTRIBUTIONTYPE,
          type
        );
        if (distributionType) {
          type = getTranslateText(distributionType.prefLabel);
        } else {
          type = null;
        }
      }
      if (format && typeof format !== 'undefined') {
        formatString = format.join(', ');
      }
      nodes.push(` ${type} (${formatString})`);
    });
    return nodes;
  };

  if (distribution && _.isArray(Object.keys(distribution))) {
    return (
      <div className="d-flex flex-wrap mb-4">
        {localization.search_hit.distributionType}:{children(distribution)}
      </div>
    );
  }
  return null;
};

const renderAccessRights = accessRight => {
  if (!accessRight) {
    return null;
  }

  const { code } = accessRight || {
    code: null
  };
  let accessRightsIconClass;
  let accessRightsColorClass;
  let accessRightsLabel;

  switch (code) {
    case 'NON_PUBLIC':
      accessRightsIconClass = 'fdk-icon-non-public';
      accessRightsColorClass = 'fdk-color-unntatt';
      accessRightsLabel =
        localization.dataset.accessRights.authorityCode.nonPublicDetailsLabel;
      break;
    case 'RESTRICTED':
      accessRightsIconClass = 'fdk-icon-restricted';
      accessRightsColorClass = 'fdk-color-restricted';
      accessRightsLabel =
        localization.dataset.accessRights.authorityCode.restrictedDetailsLabel;
      break;
    case 'PUBLIC':
      accessRightsIconClass = 'fdk-icon-public';
      accessRightsColorClass = 'fdk-color-offentlig';
      accessRightsLabel =
        localization.dataset.accessRights.authorityCode.publicDetailsLabel;
      break;
    default:
      accessRightsLabel = localization.unknown;
  }

  return (
    <div className="d-flex align-items-center mb-4">
      <i className={`mr-2 ${accessRightsIconClass}`} />
      <div className={accessRightsColorClass}>{accessRightsLabel}</div>
    </div>
  );
};

const renderSample = dataset => {
  const { sample } = dataset;
  if (Array.isArray(sample) && sample.length > 0) {
    return (
      <div className="mb-4" id="search-hit-sample">
        {localization.search_hit.sample}
      </div>
    );
  }
  return null;
};

export const SearchHitItem = props => {
  const { referenceData, dataset } = props;
  const { id, catalog, publisher, theme, provenance, accessRights } =
    dataset || {};
  let { title, description, objective } = dataset || {};
  title = getTranslateText(title);
  description = getTranslateText(description);
  objective = getTranslateText(objective);

  if (description && description.length > 220) {
    description = `${description.substr(0, 220)}...`;
  } else if (description && description.length < 150 && objective) {
    const freeLength = 200 - description.length;
    const objectiveLength = objective.length;
    description = `${description} ${objective.substr(0, 200 - freeLength)} ${
      objectiveLength > freeLength ? '...' : ''
    }`;
  }
  const link = `/datasets/${id}`;

  return (
    <article>
      <article
        className="fdk-a-search-hitx search-hit"
        title={`${localization.result.dataset}: ${title}`}
      >
        <span className="uu-invisible" aria-hidden="false">
          SÃ¸ketreff.
        </span>

        <header>
          <SearchHitHeader
            tag="h2"
            title={title}
            titleLink={link}
            catalog={catalog}
            publisherLabel={localization.search_hit.owned}
            publisher={publisher}
            theme={theme}
            nationalComponent={_.get(provenance, 'code') === 'NASJONAL'}
            referenceData={referenceData}
            darkThemeBackground={false}
          />
        </header>
        <p className="fdk-text-size-medium">
          <span className="uu-invisible" aria-hidden="false">
            Beskrivelse av datasettet,
          </span>
          {description}
        </p>

        {renderAccessRights(accessRights)}

        {renderFormats(dataset, referenceData)}

        {renderSample(dataset)}
      </article>
    </article>
  );
};

SearchHitItem.defaultProps = {
  dataset: null,
  referenceData: null
};

SearchHitItem.propTypes = {
  dataset: PropTypes.object,
  referenceData: PropTypes.object
};
