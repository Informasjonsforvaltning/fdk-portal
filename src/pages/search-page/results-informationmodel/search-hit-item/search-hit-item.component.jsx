import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';

import localization from '../../../../lib/localization';
import './search-hit-item.scss';
import { getTranslateText } from '../../../../lib/translateText';
import { SearchHitHeader } from '../../../../components/search-hit-header/search-hit-header.component';

const renderHeaderLink = (item, publisher, publishers) => {
  if (!item) {
    return null;
  }
  const { title } = item;
  const link = `/informationmodels/${encodeURIComponent(item.id)}`;

  return (
    <header>
      <SearchHitHeader
        tag="h2"
        title={getTranslateText(title)}
        titleLink={link}
        publisherLabel={`${localization.responsible}:`}
        publisher={publisher}
        publisherItems={publishers}
        nationalComponent={item.nationalComponent}
        darkThemeBackground={false}
      />
    </header>
  );
};

export const SearchHitItem = props => {
  const { item, fadeInCounter, publishers } = props;

  const searchHitClass = cx('search-hit', {
    'fade-in-200': fadeInCounter === 0,
    'fade-in-300': fadeInCounter === 1,
    'fade-in-400': fadeInCounter === 2
  });

  return (
    <article className={searchHitClass}>
      <span className="uu-invisible" aria-hidden="false">
        SÃ¸ketreff.
      </span>

      {renderHeaderLink(item, _.get(item, 'publisher'), publishers)}
    </article>
  );
};

SearchHitItem.defaultProps = {
  fadeInCounter: null,
  item: null,
  publishers: null
};

SearchHitItem.propTypes = {
  fadeInCounter: PropTypes.number,
  item: PropTypes.shape({}),
  publishers: PropTypes.object
};
