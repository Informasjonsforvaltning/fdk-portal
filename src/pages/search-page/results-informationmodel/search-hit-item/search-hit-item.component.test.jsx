import React from 'react';
import { shallow } from 'enzyme';
import { SearchHitItem } from './search-hit-item.component';
import informationModelItem from './__fixtures/informationModelResponse.json';

let defaultProps;

test('should render api-SearchHitItem correctly with no props', () => {
  const wrapper = shallow(<SearchHitItem />);
  expect(wrapper).toMatchSnapshot();
});

test('should render informatiomModel-SearchHitItem correctly', () => {
  defaultProps = {
    item: informationModelItem
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
