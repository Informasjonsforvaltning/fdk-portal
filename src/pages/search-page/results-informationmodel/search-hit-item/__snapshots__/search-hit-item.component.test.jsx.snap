// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render api-SearchHitItem correctly with no props 1`] = `
<article
  className="search-hit"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
</article>
`;

exports[`should render informatiomModel-SearchHitItem correctly 1`] = `
<article
  className="search-hit"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <header>
    <SearchHitHeader
      darkThemeBackground={false}
      externalLink={false}
      nationalComponent={false}
      publisher={
        Object {
          "id": "970018131",
          "name": "UTDANNINGSDIREKTORATET",
          "orgPath": "/STAT/872417842/970018131",
          "prefLabel": Object {
            "no": "Utdanningsdirektoratet",
          },
          "uri": "http://data.brreg.no/enhetsregisteret/enhet/970018131",
        }
      }
      publisherItems={null}
      publisherLabel="Responsible:"
      publisherTag="span"
      referenceData={null}
      statusCode={null}
      tag="h2"
      theme={Array []}
      title="Enhetsregisteret"
      titleLink="/informationmodels/eadda99c-781e-4900-8f54-8dd298168382"
    />
  </header>
</article>
`;
